jsettlers.logic/src/main/java/jsettlers/logic/buildings/workers/WorkerAnimationBuilding.java

```diff
@@ -0,0 +1,140 @@
+	private static final long serialVersionUID = 1210188934302286757L;
+
+	private int[] animationStartTime;
+	private boolean[] isAnimationRequested;
+
+	protected WorkerAnimationBuilding(EBuildingType type, Player player, ShortPoint2D position, IBuildingsGrid buildingsGrid) {
+		super(type, player, position, buildingsGrid);
+
+		animationStartTime = new int[getAnimationCount()];
+		isAnimationRequested = new boolean[getAnimationCount()];
+
+		for (int i = 0; i < getAnimationCount(); i++) {
+			isAnimationRequested[i] = false;
+			animationStartTime[i] = 0;
+		}
+	}
+
+	protected abstract int[] getAnimationDurations();
+	protected abstract int[] getAnimationTransitions();
+
+
+	public int getAnimationCount() {
+		if (getAnimationDurations() != null) {
+			return getAnimationDurations().length;
+		} else {
+			return 0;
+		}
+	}
+
+	private int getTransitionCount() {
+		if (getAnimationTransitions() != null) {
+			return getAnimationTransitions().length;
+		} else {
+			return 0;
+		}
+	}
+
+	public boolean isAnimationRequested(int index) {
+		if (index >= 0 && index < getAnimationCount()) {
+			return isAnimationRequested[index];
+		}
+
+		return false;
+	}
+
+	public float getAnimationProgress(int index) {
+		if (index >= 0 && index < getAnimationCount()) {
+			if (isAnimationRequested[index]) {
+				float progress = (MatchConstants.clock().getTime() - animationStartTime[index]) / (float)getAnimationDurations()[index];
+
+				//check transition for possible next animation
+				if (!isAnimationRequested(index + 1) && index < getTransitionCount()) {
+					float nextAnimationQuotient = (float)getAnimationTransitions()[index] / getAnimationDurations()[index];
+
+					if (progress >= nextAnimationQuotient) {
+						requestAnimation(index + 1);
+					}
+				}
+
+				if (progress < 1f) {
+					return progress;
+				} else {
+					isAnimationRequested[index] = false;
+					return 1f;
+				}
+			}
+		}
+
+		return 0f;
+	}
+
+	/**
+	 * Invokes the animation for the building
+	 *
+	 * @return the duration of the animation which is requested by index
+	 */
+	public int requestAnimation(int index) {
+		if (index >= 0 && index < getAnimationCount()) {
+			if (!isAnimationRequested[index]) {
+				isAnimationRequested[index] = true;
+				animationStartTime[index] = MatchConstants.clock().getTime();
+				return getAnimationDurations()[index];
+			}
+		}
+
+		return 0;
+	}
+
+	/**
+	 * Requests all animations
+	 * @return entire duration of all animations
+	 */
+	public int requestAnimation() {
+		requestAnimation(0);
+		return getEntireDuration();
+	}
+
+	public int getEntireDuration() {
+		if (getAnimationDurations() != null) {
+			int transitionSum = 0;
+			for (int i = 0; i < getTransitionCount(); i++) {
+				transitionSum += getAnimationDurations()[i] - getAnimationTransitions()[i];
+			}
+
+			return Arrays.stream(getAnimationDurations()).sum() - transitionSum;
+		}
+		return 0;
+	}
+}
```

Comment:
This is very bloated and introduces many small and redundant classes. I think there should just be a flag (isPlayingAnimation) and a progress function and the graphics code matches these with specific animations